Add Process support for UEFI

Ayush1325 opened the pull request:
UEFI does not have an actual process. However, it does provide methods to launch and execute another UEFI image. Having process support is important since it is possible to run rust test suit using `Command::output` and is the first step towards being able to run it for UEFI.

Here is an overview of how the support is implemented.

- We create a copy of the SystemTable. This is required since at least OVMF seems to crash if the original system table is modified.
- Stdout and Stderr pipe works by registering a new `simple_text_output` Protocol and pointing the child system table to use those.
- `Stdio::Inherit` just points the console to the current running image console which seems to work with even 3 levels of process.
- `spawn` is left unimplemented since it does not make sense for UEFI architecture. Additionally, since https://github.com/rust-lang/rust/pull/105458 was merged, the `spawn` and `output` implementations are completely independent.

Ayush1325 committed:
uefi: Add process

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 committed:
uefi: process: Add support to capture stdout

Use a custom simple_text_output protocol to capture output.

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 committed:
uefi: process: Add stderr support

Implement stderr support in similar fashion.

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 committed:
uefi: process: Add null protocol

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 committed:
uefi: process Implement inherit

Only tested in 2 levels right now. Need args support for 3 levels

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 committed:
uefi: process: Add support for args

Also fix stdio inherit

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 committed:
uefi: process: Add CommandArgs support

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 committed:
uefi: process: Final Touchups

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Nilstrieb commented:
@dvdhrm @nicholasbishop could I get a review from you?

Nilstrieb commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,665 @@
+    pub stderr: Option<AnonPipe>,
+}
+
+// FIXME: This should be a unit struct, so we can always construct it
```
What's stopping you from doing this in this PR?

Ayush1325 replied:
I will remove that comment. It's a remnant from `unsupported/process.rs`. It is actually possible to use `Stdio::Inherit`, `Stdio::Null`, `Stdio::MakePipe` in UEFI.

joboet commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,663 @@
+        }
+
+        pub fn stdout_init(&mut self, mut protocol: helpers::Protocol<PipeProtocol>) {
+            self.st.console_out_handle = protocol.handle().as_ptr();
```
I think these updates need to adjust the CRC in the table header, otherwise very paranoid images will complain.

Ayush1325 committed:
uefi: process: Fixes from PR

- Update system table crc32
- Fix unsound use of Box
- Free exit data
- Code improvements
- Introduce OwnedTable
- Update r-efi to latest version
- Use extended_varargs_abi_support for
  install_multiple_protocol_interfaces and
  uninstall_multiple_protocol_interfaces
- Fix comments
- Stub out args implementation

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

joboet commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,686 @@
+
+                let loaded_image: NonNull<loaded_image::Protocol> =
+                    helpers::open_protocol(child_handle, loaded_image::PROTOCOL_GUID).unwrap();
+                let mut st: Box<r_efi::efi::SystemTable> =
+                    Box::new(unsafe { crate::ptr::read((*loaded_image.as_ptr()).system_table) });
```
To preserve forward-compatibility if new fields are added to the system table, I think we should copy `system_table.hdr.header_size` bytes instead of the fields defined  in `r_efi`.

Ayush1325 replied:
How do you propose going about this? For the compiler, `T` is `Sized`, but we want to construct and use it as if `T` is `!Sized`. So I am not quite sure how to proceed. Maybe create an `OwnedTable`? It doesn't seem like I can use standard Box for this.

Ayush1325 replied:
I have added an `OwnedTable` type

joboet commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,686 @@
+                clear_screen: Self::clear_screen,
+                set_cursor_position: Self::set_cursor_position,
+                enable_cursor: Self::enable_cursor,
+                mode: mode.as_mut(),
```
This isn't sound because `Box` is `noalias`, so this reference will be invalidated once `mode` is moved
```suggestion
                mode: Box::into_raw(mode),
```
and add a `Drop` implementation that calls `Box::from_raw` to deallocate.

joboet commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,686 @@
+                clear_screen: Self::clear_screen,
+                set_cursor_position: Self::set_cursor_position,
+                enable_cursor: Self::enable_cursor,
+                mode: mode.as_mut(),
```
Likewise.

joboet commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,686 @@
+                    Box::new(unsafe { crate::ptr::read((*loaded_image.as_ptr()).system_table) });
+
+                unsafe {
+                    (*loaded_image.as_ptr()).system_table = st.as_mut();
```
As described below, this isn't sound, as this reference will be invalidated as soon as `st` is returned.

joboet commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,686 @@
+
+            // Drop exitdata
+            unsafe {
+                exit_data_size.assume_init_drop();
+                exit_data.assume_init_drop();
```
This doesn't do anything, `usize` and `*mut u16` don't implement `Drop`. We should call the allocator instead.

Ayush1325 replied:
Wouldn't MaybeUninit<usize> be allocated on stack? Only exit data is allocated by the other image we launch.

joboet reviewed (changes requested):
There are a number of cases where we open a protocol, but never close it. Maybe a `BorrowedProtocol` wrapper with the necessary `Drop` implementation could help solve this?

Nilstrieb commented:
r? joboet

Ayush1325 commented:
> There are a number of cases where we open a protocol, but never close it. Maybe a BorrowedProtocol wrapper with the necessary Drop implementation could help solve this?

We are using `GET_PROTOCOL` attribute to open protocol, which does not require us to close the protocol. Feel free to point out places that might benefit from open protocol in a different way since those will require BorrowedProtocol wrapper.

joboet reviewed (commented):
I think this PR could be simplified quite a bit, which would also help with soundness:

Currently, `Command` in `uefi_command_internal` stores things like the system table, stdout, stderr, but it also references these resources in the image load options, or the system table. So in effect, we have a self-referential structure. Those are really tricky to get right, and this PR has some soundness bugs related to them.

Luckily, we can morph this into a Rust-like form, because all these things only need to live for the duration of the `output` call. So I'd recommend just storing these resources in variables inside `output`, and then referencing them at the appropriate places:
```rust
fn output(&mut self) {
    let stdout = create_pipe(...);
    let stderr = create_pipe(...);
    let args = convert_args(...);
    let mut st = create_st(&stdout, &stderr);
    run(&args, &mut st);
}
```

@rustbot author

Dylan-DPC commented:
@Ayush1325 any updates on this? thanks

Ayush1325 commented:
> @Ayush1325 any updates on this? thanks

Hi. Sorry for the delay. My college exams were going on and am busy with some contract work. Will try to address everything in a few days

Ayush1325 commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,699 @@
+    type Item = &'a OsStr;
+
+    fn next(&mut self) -> Option<&'a OsStr> {
+        self.iter.next().map(|x| x.as_ref())
```
This does not seem to work:
```shell
error[E0308]: mismatched types
   --> library/std/src/sys/pal/uefi/process.rs:305:9
    |
304 |     fn next(&mut self) -> Option<&'a OsStr> {
    |                           ----------------- expected `core::option::Option<&'a OsStr>` because of return type
305 |         self.iter.next().as_deref()
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<&OsStr>`, found `Option<&OsString>`
    |
    = note: expected enum `core::option::Option<&'a OsStr>`
               found enum `core::option::Option<&OsString>`

For more information about this err
```

joboet replied:
Ah right. Nevermind.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8957851947/job/24601405539) [(plain)](https://github.com/rust-lang/rust/commit/8898c747464420c44c009c843fec1dd7d1e4d391/checks/24601405539/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/69ffc0d3a3c619009bcb27b8f61d810e27b12612/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
curl: (22) The requested URL returned error: 404
ERROR: failed to download llvm from ci

    HELP: There could be two reasons behind this:
        1) The host triple is not supported for `download-ci-llvm`.
        2) Old builds get deleted after a certain time.
    HELP: In either case, disable `download-ci-llvm` in your config.toml:
    [llvm]
    download-ci-llvm = false
    
Build completed unsuccessfully in 0:00:01
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8957890857/job/24601494517) [(plain)](https://github.com/rust-lang/rust/commit/cbe87cc05dfc72b606e02738180912c635f05a10/checks/24601494517/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/69ffc0d3a3c619009bcb27b8f61d810e27b12612/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
curl: (22) The requested URL returned error: 404
ERROR: failed to download llvm from ci

    HELP: There could be two reasons behind this:
        1) The host triple is not supported for `download-ci-llvm`.
        2) Old builds get deleted after a certain time.
    HELP: In either case, disable `download-ci-llvm` in your config.toml:
    [llvm]
    download-ci-llvm = false
    
Build completed unsuccessfully in 0:00:00
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #124784) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Ayush1325 commented on library/std/src/sys/pal/uefi/helpers.rs:
```diff
@@ -221,3 +221,145 @@ pub(crate) fn runtime_services() -> Option<NonNull<r_efi::efi::RuntimeServices>>
+        if let Some(bt) = boot_services() {
+            let bt: NonNull<r_efi::efi::BootServices> = bt.cast();
+            unsafe {
+                ((*bt.as_ptr()).uninstall_protocol_interface)(
```
Will switch to `uninstall_multiple_protocol_interfaces` once https://github.com/r-efi/r-efi/pull/79 is merged.

dvdhrm replied:
I dont think varargs are supported for eficall. I think it is only supported for "extern C". So the multi-call is not an option.

Ayush1325 replied:
Yes, but it seems `uninstall_multiple_protocol_interfaces` (and `install_multiple_protocol_interfaces`) are to be preferred over their single counterparts due to EFI 1.1 extensions that allow tracking. Thus `uninstall_multiple_protocol_interfaces` only fails if `One of the protocol interfaces was not previously installed on Handle. `, at least according to spec.

dvdhrm replied:
What I am saying is that `uninstall_multiple_protocol_interfaces` uses a variable number of arguments, which is not supported by stable Rust for `eficall`. The prototype provided by `r-efi` is not usable in its current form and can easily lead to ABI violations.

You can look at EDK2 and see how `uninstall_multiple_protocol_interfaces` is implemented, and simply copy the logic from there.

dvdhrm replied:
Also see: https://github.com/rust-lang/rust/issues/100189

Since `std` can use unstable features, you can just provide your own `fn` type and transmute the fn-pointer from the boot-services table.

Ayush1325 commented:
@rustbot ready

joboet reviewed (commented):
Sorry for the delay, haven't had that much time recently. It's looking quite good and sound, but some smaller problems remain.

Ayush1325 commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,712 @@
+    pub const SUCCESS: ExitCode = ExitCode(false);
+    pub const FAILURE: ExitCode = ExitCode(true);
+
+    pub fn as_i32(&self) -> i32 {
+        self.0 as i32
+    }
```
Yes, it should be done, but this is probably not the PR to do it.

Ayush1325 commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,712 @@
+                acc.push(arg);
+                acc
+            });
+            cmd.set_args(&args);
```
Do you mean I should extract argument parsing to it's own function and leave the current logic, or just have a blank implementation for handling arguments for now?

Also, what are the rules regarding args?
1. Is the user supposed to put `"` around string arguments (those that contain whitespace). Or is the implementation supposed to detect if quote is needed?
2. Is the user supposed to escape the characters using `^` (UEFI escape), or should the implementation detect characters that need to be escaped?
3. What happens if user passes a NULL character?

joboet replied:
We need to do the escaping so that `env::args()` in the image can recover the exact arguments passed in here. As for NULL, UNIX remembers whether one of the arguments contains NULL inside the `Command` structure and returns an error upon spawning in that case.

joboet commented on library/std/src/sys/pal/uefi/helpers.rs:
```diff
@@ -221,3 +227,192 @@ pub(crate) fn runtime_services() -> Option<NonNull<r_efi::efi::RuntimeServices>>
+    }
+}
+
+impl AsRef<r_efi::protocols::device_path::Protocol> for DevicePath {
+    fn as_ref(&self) -> &r_efi::protocols::device_path::Protocol {
+        unsafe { self.0.as_ref() }
+    }
+}
+
+impl AsMut<r_efi::protocols::device_path::Protocol> for DevicePath {
+    fn as_mut(&mut self) -> &mut r_efi::protocols::device_path::Protocol {
+        unsafe { self.0.as_mut() }
+    }
+}
```
The protocol is a variable-width structure, but this returns a reference to a fixed-width node. IIRC, it's still unclear whether taking the reference restricts the provenance of the reference to the reference type, but until that is figured out, I'd prefer if this were expressed as an `as_ptr()` or `as_mut_ptr()` method instead.

joboet commented on library/std/src/sys/pal/uefi/helpers.rs:
```diff
@@ -221,3 +227,192 @@ pub(crate) fn runtime_services() -> Option<NonNull<r_efi::efi::RuntimeServices>>
+        };
+
+        if r.is_error() {
+            return Err(crate::io::Error::from_raw_os_error(r.as_usize()));
```
We need to drop the allocated protocol here.
```suggestion
            drop(Box::from_raw(protocol));
            return Err(crate::io::Error::from_raw_os_error(r.as_usize()));
```

joboet commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,706 @@
+    }
+
+    pub fn arg(&mut self, arg: &OsStr) {
+        self.args.push(arg.to_os_string());
```
Could you, for the moment, stub this out? This will not behave as expected.
```suggestion
        panic!("unsupported")
```

joboet commented on library/std/src/sys/pal/uefi/process.rs:
```diff
@@ -0,0 +1,706 @@
+
+                let loaded_image: NonNull<loaded_image::Protocol> =
+                    helpers::open_protocol(child_handle, loaded_image::PROTOCOL_GUID).unwrap();
+                let st: helpers::OwnedTable<r_efi::efi::SystemTable> =
```
```suggestion
                let st =
```

Ayush1325 commented:
@rustbot ready

joboet commented on library/std/src/sys/pal/uefi/helpers.rs:
```diff
@@ -221,3 +227,185 @@ pub(crate) fn runtime_services() -> Option<NonNull<r_efi::efi::RuntimeServices>>
+            p: &OsStr,
+            protocol: NonNull<r_efi::protocols::device_path_from_text::Protocol>,
+        ) -> io::Result<DevicePath> {
+            let path_vec = p.encode_wide().chain(Some(0)).collect::<Vec<u16>>();
```
This will ignore internal NULL bytes in the string, resulting in truncation.

Ayush1325 replied:
I am not sure I understand. Can you provide me with an example?

joboet replied:
Exactly, yes. We should return an error in that case instead, [like on Windows](https://github.com/rust-lang/rust/blob/5ced3dad5752cc9423a19e9a5aafad6e12ed3063/library/std/src/sys/pal/windows/mod.rs#L177-L196).

Ayush1325 replied:
I am not doing manual loop unrolling like the windows version since I am not sure if it is necessary.

workingjubilee replied:
don't worry. the loop unrolling was from quite some time ago. we should test to see if it even does anything anymore.

joboet reviewed (commented):
Just one bug left :tada:. I'll be filing an ACP for introducing a `RawExitStatus` and wait for it to be discussed on Tuesday before merging this, just so we are not locking in any choices, but otherwise this seems good to go.

joboet reviewed (approved):
I think this is ready to go. Let's wait until https://github.com/rust-lang/libs-team/issues/403 is resolved (there was no libs-api meeting yesterday) and then merge this.

joboet commented:
Thank you and sorry for the delay!
@bors r+

bors commented:
:pushpin: Commit 4a868f41520c20553869299b7735025b1ddd7e71 has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 4a868f41520c20553869299b7735025b1ddd7e71 -->
<!-- homu: {"type":"Approved","sha":"4a868f41520c20553869299b7735025b1ddd7e71","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
This failed the rollup because we are starting to forbid `unsafe_op_in_unsafe_fn`, https://github.com/rust-lang/rust/pull/127954#issuecomment-2238676554. It looks like this just misses a few cases, you'll probably start getting the warnings locally if you rebase.

@bors r-

workingjubilee commented:
This looks basically like how I last saw it, yeah.

@bors r=joboet

bors commented:
:pushpin: Commit e2903989da89b43722efca49bcc21749014739ff has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet e2903989da89b43722efca49bcc21749014739ff -->
<!-- homu: {"type":"Approved","sha":"e2903989da89b43722efca49bcc21749014739ff","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Ayush1325 commented:
> This looks basically like how I last saw it, yeah.
> 
> @bors r=joboet

Yup, there was only 1 place of failure after rebase: https://github.com/rust-lang/rust/pull/123196/files#diff-21f19573f63f3900715f66b8ad29e271024c92279c5d73e0b851af423d213129R328
